src/middlewared/middlewared/plugins/tunables.py

```diff
@@ -25,212 +26,110 @@ class TunableService(CRUDService):
         Update Tunable of `id`.
         """
-        old = await self.get_instance(id)
+        old = await self.get_instance(_id)
 
         new = old.copy()
         new.update(data)
+        if old == new:
+            # nothing updated so return early
+            return old
 
-        await self.clean(new, 'tunable_update', old=old)
-        await self.validate(new, 'tunable_update')
+        if (comment := data.get('comment', '').strip()) and comment != new['comment']:
+            new['comment'] = comment
 
-        await self.lower(new)
+        if not new['enabled'] and new['orig_value']:
+            # before we delete from db, let's set the tunable back to it's original value
+            await self.middleware.run_in_thread(self.set_original_value, new['var'], new['orig_value'])
 
-        await self.middleware.call(
-            'datastore.update',
-            self._config.datastore,
-            id,
-            new,
-            {'prefix': self._config.datastore_prefix}
+        _id = await self.middleware.call(
+            'datastore.update', self._config.datastore, _id, new, {'prefix': self._config.datastore_prefix}
         )
+        await self.middleware.call('service.restart', 'sysctl')
+        return await self.get_instance(_id)
 
-        if old['type'] == 'SYSCTL' and old['var'] in self.__default_sysctl and (
-            old['var'] != new['var'] or old['type'] != new['type']
-        ):
-            default_value = self.__default_sysctl.pop(old['var'])
-            cp = await run(['sysctl', f'{old["var"]}={default_value}'], check=False, encoding='utf8')
-            if cp.returncode:
-                self.middleware.logger.error(
-                    'Failed to set sysctl %r -> %r : %s', old['var'], default_value, cp.stderr
-                )
-
-        await self.middleware.call('service.reload', new['type'])
-
-        return await self.get_instance(id)
-
-    async def do_delete(self, id):
+    async def do_delete(self, _id):
         """
         Delete Tunable of `id`.
         """
-        tunable = await self.get_instance(id)
-        await self.lower(tunable)
-        if tunable['type'] == 'sysctl':
-            # Restore the default value, if it is possible.
-            value_default = self.__default_sysctl.pop(tunable['var'], None)
-            if value_default:
-                cp = await run(['sysctl', f'{tunable["var"]}={value_default}'], check=False, encoding='utf8')
-                if cp.returncode:
-                    self.middleware.logger.error(
-                        'Failed to set sysctl %r -> %r : %s', tunable['var'], value_default, cp.stderr
-                    )
-
-        response = await self.middleware.call(
-            'datastore.delete',
-            self._config.datastore,
-            id
-        )
-
-        await self.middleware.call('service.reload', tunable['type'].lower())
-
-        return response
+        _id = await self.get_instance(_id)
+        if _id['orig_value']:
```

Comment:
I am torn on what to do here, we have explicit check in creation that if we are not able to determine original value - we are not going to allow that tunable to be created which implies that we should already have this value. Having this check means that we just have it for already configured tunables. I am not sure what might be the best approach here, would be really nice if we could somehow normalise the database to avoid this maybe by running another migration and determining the original value but haven't seen if the changed values will be applied already when we run the migration

Reply:
Yeah, I thought about this too. I'm not positive that when `middlewared` service starts, that the system sysctl values are at their original value by the time that service starts. In theory they should be because I could change it so that `systemd-tunable` service runs after `middlewared`. Not sure, I'll have to test

Reply:
After testing when `ix-update.service` runs, the kernel hasn't applied any user configured sysctl tunables so I can normalize the database on upgrade 😄